executable-monitor/executable-monitor.py

```diff
@@ -0,0 +1,140 @@
+    parser.add_argument('--number-of-runs',
+                        type=int,
+                        required=True,
+                        help='Timeout for each executable run.')
+    parser.add_argument('--success-line',
+                        type=str,
+                        required=False,
+                        help='Line that indicates executable completed successfully. Required if --success-exit-status is not used.')
+    parser.add_argument('--success-exit-status',
+                        type=int,
+                        required=False,
+                        help='Exit status that indicates that the executable completed successfully. Required if --success-line is not used.')
+
+    args = parser.parse_args()
+
+    if args.success_exit_status is None and args.success_line is None:
+        print("Must specify at least one of the following: --success-line, --success-exit-status.")
+        sys.exit(1)
+
+    if not os.path.exists(args.exe_path):
+        print(f'Input executable path \"{args.exe_path}\" does not exist.')
+        sys.exit(1)
+
+    # Create log directory if it does not exist.
+    if not os.path.exists(args.log_dir):
+        os.makedirs(args.log_dir, exist_ok = True)
+
+    # Convert any relative path (like './') in passed argument to absolute path.
+    EXE_PATH = os.path.abspath(args.exe_path)
+    LOG_DIR = os.path.abspath(args.log_dir)
+
+    print(f"Running executable: {EXE_PATH} ")
+    print(f"Storing logs in: {LOG_DIR}")
+    print(f"Timeout per run (seconds): {args.timeout_seconds}")
+    print(f"Number of runs: {args.number_of_runs}\n")
+
+    EXE_NAME = os.path.basename(EXE_PATH)
+
+    log_file = open(f'{LOG_DIR}/{EXE_NAME}_output.txt', 'w')
+
+    exe = subprocess.Popen([EXE_PATH], stdout=subprocess.PIPE, stderr=subprocess.STDOUT, universal_newlines=True)
+  
+    cur_time_seconds = time.time()
+    timeout_time_seconds = cur_time_seconds + args.timeout_seconds
+    timeout_occurred = False
+
+    exe_exit_status = None
+    exe_exitted = False
+
+    success_line_found = False
+    cur_line_ouput = 1
+
+    wait_for_exit = args.success_exit_status is not None       
+
+    sys.stdout.write("DEVICE OUTPUT:\n\n")
+    log_file.write("DEVICE OUTPUT:\n\n")
+
+    while not (timeout_occurred or exe_exitted or (not wait_for_exit and success_line_found)):
+
+        # Read executable's stdout and write to stdout and logfile
+        exe_stdout_line = exe.stdout.readline()
+        sys.stdout.write(exe_stdout_line)
+        log_file.write(exe_stdout_line)
+
+        # Check if the executable printed out it's success line
+        if args.success_line is not None and args.success_line in exe_stdout_line:
+            success_line_found = True
+
+        # Check if executable exitted
+        exe_exit_status = exe.poll()
+        if exe_exit_status is not None:
+            exe_exitted = True
+
+        # Check for timeout
+        cur_time_seconds = time.time()
```

Comment:
Unironically, since we have a time based check happening, it might be worth putting a sleep(1/2/3/4/5) in here? Can see a situation where we run through this loop a lot before even hitting a few seconds on the demo.

Reply:
Note that I read output line by line to detect if the success line is hit while in this loop so I cannot do this

Reply:
Yeah that "readlink blocks until it gets a full line" is exactly what I was worried about.
subprocess starts -> doesn't write a full line since it dies/hangs/something goes wrong -> sit at readline() forever since it blocks -> have to wait until the github action times out and marks this as a failure

Can make the argument that the above would be a hard to hit edge case.  Just something I was thinking about.